ledger/trie/light_node.go

```diff
@@ -0,0 +1,316 @@
+// Copyright 2021 Optakt Labs OÜ
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package trie
+
+import (
+	"encoding/hex"
+	"fmt"
+	"io"
+
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/ledger/common/encoding"
+	"github.com/onflow/flow-go/ledger/common/hash"
+	"github.com/onflow/flow-go/ledger/common/utils"
+	"github.com/optakt/flow-dps/models/dps"
+)
+
+type LightNode struct {
+	LIndex uint64
+	RIndex uint64
+
+	Path      []byte
+	HashValue []byte
+
+	// Height where the node is at.
+	Height uint16
+	// Height at which the node skips if it is an extension.
+	Skip uint16
+}
+
+type IndexMap map[Node]uint64
+
+func ToLightNode(node Node, index IndexMap) (*LightNode, error) {
+	leftIndex, found := index[node.LeftChild()]
+	if !found {
+		hash := node.LeftChild().Hash()
+		return nil, fmt.Errorf("missing node with hash %s", hex.EncodeToString(hash[:]))
+	}
+	rightIndex, found := index[node.RightChild()]
+	if !found {
+		hash := node.RightChild().Hash()
+		return nil, fmt.Errorf("missing node with hash %s", hex.EncodeToString(hash[:]))
+	}
+
+	// By calling node.Hash we ensure that the node hash is computed and will not be stored dirty.
+	hash := node.Hash()
```

Comment:
It is necessary unfortunately, because `node.Hash` returns a `ledger.Hash` which is an alias to a `[32]byte`, which can't be used as a `[]byte` unless we slice it by using `[:]`. This however, can not be done on the return of a function, so we need to store it in a variable:

```go
node.Hash()[:] <- does not compile

tmp := node.Hash()
slice := tmp[:] <- compiles, slice is a slice and not an array
```

Reply:
Got it